File: linked_list/linked_list.py
#Asilbek Asqarov
#Mohirdev Maʻlumotlar Tuzilmasi va Algoritmlar darslari
#dars-03
#Linked List

class Node:

    def __init__(self, data):
        self.data = data
        self.next = None
    
class Linked_list:

    def __init__(self):
        self.head = None

    def pprint(self):
        temp = self.head

        while temp:
            print(temp.data)
            temp = temp.next
        
    def push(self, new_data):
        
        new_node = Node(new_data)
        
        new_node.next = self.head
        self.head = new_node
    
    def add_after(self, prev_node, new_data):
        new_node = Node(new_data)
        if prev_node is None:
            print("No node!")
            return 
        
        new_node.next = prev_node.next
        prev_node.next = new_node
    
    def append(self, new_data):
        new_node = Node(new_data)
        if self.head is None:
            self.head = new_node
            return
        
        temp =self.head
        
        while temp.next:
            temp = temp.next
        
        temp.next = new_node 
    
    def delete(self, key):
        
        temp = self.head
        
        if temp and temp.data == key:
            self.head = temp.next 
            temp = None
            return 
        
        while temp:
            
            if temp.data == key:
                break
            prev = temp
            temp = temp.next
        
        if temp==None:
            return
        prev.next = temp.next
        temp = None
        
    
        
        
        
        
        
        
        
        
        
                

File: linked_list/linked_list.cpp
#include <iostream>
using namespace std;

struct Node
{
    int data;
    Node *next;

    Node(int val)
    {
        data = val;
        next = nullptr;
    }
};

class LinkedList
{
public:
    Node *head;

    LinkedList() { head = nullptr; }

    void pprint()
    {
        Node *temp = head;
        while (temp != nullptr)
        {
            cout << temp->data << endl;
            temp = temp->next;
        }
    }

    void push(int new_data)
    {
        Node *new_node = new Node(new_data);
        new_node->next = head;
        head = new_node;
    }

    void add_after(Node *prev_node, int new_data)
    {
        if (prev_node == nullptr)
        {
            cout << "No node!" << endl;
            return;
        }
        Node *new_node = new Node(new_data);
        new_node->next = prev_node->next;
        prev_node->next = new_node;
    }

    void append(int new_data)
    {
        Node *new_node = new Node(new_data);
        if (head == nullptr)
        {
            head = new_node;
            return;
        }
        Node *temp = head;
        while (temp->next != nullptr)
        {
            temp = temp->next;
        }
        temp->next = new_node;
    }

    void deleteNode(int key)
    {
        Node *temp = head;
        Node *prev = nullptr;

        if (temp != nullptr && temp->data == key)
        {
            head = temp->next;
            delete temp;
            return;
        }

        while (temp != nullptr && temp->data != key)
        {
            prev = temp;
            temp = temp->next;
        }

        if (temp == nullptr)
            return;

        prev->next = temp->next;
        delete temp;
    }
};

int main()
{
    LinkedList ll;
    ll.append(10);
    ll.append(12);
    ll.append(19);
    ll.pprint();
    ll.push(5);
    ll.pprint();
    ll.append(20);
    ll.pprint();
    ll.add_after(ll.head, 7);
    ll.pprint();

    ll.deleteNode(10);

    ll.pprint();
    return 0;
}